src/IO/External/InterpolateFromCocal.cpp

```diff
@@ -0,0 +1,204 @@
+                double* extrinsic_curvature_xy, double* extrinsic_curvature_xz,
+                double* extrinsic_curvature_yy, double* extrinsic_curvature_yz,
+                double* extrinsic_curvature_zz, double* rest_mass_density,
+                double* spatial_velocity_x, double* spatial_velocity_y,
+                double* spatial_velocity_z, double* pressure,
+                double* specific_internal_energy);
+}
+
+namespace io {
+
+namespace {
+DataVector to_datavector(std::vector<double> vec) {
+  DataVector result(vec.size());
+  std::copy(vec.begin(), vec.end(), result.begin());
+  return result;
+}
+}  // namespace
+
+tuples::tagged_tuple_from_typelist<cocal_tags> interpolate_from_cocal(
+    const gsl::not_null<std::mutex*> cocal_lock, const CocalIdType id_type,
+    const std::string& data_directory,
+    const tnsr::I<DataVector, 3, Frame::Inertial>& x) {
+  tuples::tagged_tuple_from_typelist<cocal_tags> result{};
+  const std::lock_guard lock{*cocal_lock};
+  const ScopedFpeState disable_fpes(false);  // test for fpe issues
+  const size_t num_points = get<0>(x).size();
+  std::vector<double> x_coords(num_points);
+  std::vector<double> y_coords(num_points);
+  std::vector<double> z_coords(num_points);
+  for (size_t i = 0; i < num_points; i++) {
+    x_coords[i] = x.get(0)[i];
+    y_coords[i] = x.get(1)[i];
+    z_coords[i] = x.get(2)[i];
+  }
+
+  std::vector<double> lapse(num_points);
+  std::vector<double> shift_x(num_points);
+  std::vector<double> shift_y(num_points);
+  std::vector<double> shift_z(num_points);
+  std::vector<double> spatial_metric_xx(num_points);
+  std::vector<double> spatial_metric_xy(num_points);
+  std::vector<double> spatial_metric_xz(num_points);
+  std::vector<double> spatial_metric_yy(num_points);
+  std::vector<double> spatial_metric_yz(num_points);
+  std::vector<double> spatial_metric_zz(num_points);
+  std::vector<double> extrinsic_curvature_xx(num_points);
+  std::vector<double> extrinsic_curvature_xy(num_points);
+  std::vector<double> extrinsic_curvature_xz(num_points);
+  std::vector<double> extrinsic_curvature_yy(num_points);
+  std::vector<double> extrinsic_curvature_yz(num_points);
+  std::vector<double> extrinsic_curvature_zz(num_points);
+  std::vector<double> rest_mass_density(num_points);
+  std::vector<double> spatial_velocity_x(num_points);
+  std::vector<double> spatial_velocity_y(num_points);
+  std::vector<double> spatial_velocity_z(num_points);
+  std::vector<double> pressure(num_points);
+  std::vector<double> specific_internal_energy(num_points);
+
+  // Pass the string and its length to Fortran
+  const char* dir_cstr =
+      data_directory.c_str();  // Convert std::string to C string
+  int len_dir = static_cast<int>(
+      data_directory.length());  // Get the length of the string
+
+  if (id_type == CocalIdType::Co) {
+    coc2cac_co(num_points, x_coords.data(), y_coords.data(), z_coords.data(),
```

Comment:
since you break this down to a plain C pointer anyway, why not use the `DataVector`'s `data()` member (inherited from Blaze https://bitbucket.org/blaze-lib/blaze/wiki/Vector%20Operations#!data-data)  instead?